File: budgetease_flutter/lib/services/behavioral_profiler.dart
import 'dart:math';
import '../models/transaction.dart';
import '../models/behavioral_profile.dart';
import 'database_service.dart';
import 'advisor_service.dart';

class BehavioralProfiler {
  /// Construit le profil comportemental basé sur les 30 derniers jours
  static BehavioralProfile buildProfile() {
    final last30Days = DateTime.now().subtract(const Duration(days: 30));
    final recentTransactions = DatabaseService.transactions.values
        .where((t) => t.date.isAfter(last30Days))
        .toList();

    // Calcul de la fréquence de dépenses
    final frequency = recentTransactions.length / 30;

    // Construction du pattern horaire
    final hourlyPattern = <int, int>{};
    for (var t in recentTransactions) {
      final hour = t.date.hour;
      hourlyPattern[hour] = (hourlyPattern[hour] ?? 0) + 1;
    }

    // Analyse des dépassements
    int overruns = 0;
    double totalOverrun = 0;

    for (int i = 0; i < 30; i++) {
      final day = DateTime.now().subtract(Duration(days: i));
      final dailyCap = AdvisorService.getRecommendedDailyCap();
      final spent = _getDailySpent(day);

      if (spent > dailyCap && dailyCap > 0) {
        overruns++;
        totalOverrun += (spent - dailyCap);
      }
    }

    return BehavioralProfile(
      spendingFrequency: frequency,
      hourlyPattern: hourlyPattern,
      overrunCount: overruns,
      averageOverrun: overruns > 0 ? totalOverrun / overruns : 0,
      lastUpdated: DateTime.now(),
    );
  }

  /// Récupère ou crée le profil stocké
  static BehavioralProfile getOrCreateProfile() {
    final box = DatabaseService.behavioralProfiles;
    
    if (box.isEmpty) {
      final profile = buildProfile();
      box.add(profile);
      return profile;
    }

    final profile = box.values.first;
    
    // Mettre à jour si plus de 24h
    if (DateTime.now().difference(profile.lastUpdated).inHours > 24) {
      final updated = buildProfile();
      profile.spendingFrequency = updated.spendingFrequency;
      profile.hourlyPattern = updated.hourlyPattern;
      profile.overrunCount = updated.overrunCount;
      profile.averageOverrun = updated.averageOverrun;
      profile.lastUpdated = updated.lastUpdated;
      profile.save();
    }

    return profile;
  }

  /// Calcule les dépenses d'une journée spécifique
  static double _getDailySpent(DateTime day) {
    final startOfDay = DateTime(day.year, day.month, day.day);
    final endOfDay = startOfDay.add(const Duration(days: 1));

    return DatabaseService.transactions.values
        .where((t) =>
            t.type == 'expense' &&
            t.date.isAfter(startOfDay) &&
            t.date.isBefore(endOfDay))
        .fold(0.0, (sum, t) => sum + t.amount);
  }

  /// Détermine si l'utilisateur a tendance à dépenser le soir
  static bool isEveningSpender() {
    final profile = getOrCreateProfile();
    final eveningTransactions = (profile.hourlyPattern[18] ?? 0) +
        (profile.hourlyPattern[19] ?? 0) +
        (profile.hourlyPattern[20] ?? 0) +
        (profile.hourlyPattern[21] ?? 0);
    
    final totalTransactions = profile.hourlyPattern.values.fold(0, (a, b) => a + b);
    
    return totalTransactions > 0 && (eveningTransactions / totalTransactions) > 0.4;
  }
}

File: src/lib/behavioral-profiler.ts
import { db } from './db';
import { Transaction, BehavioralProfile } from '../types';
import { getRecommendedDailyCap } from './calculations';

const PROFILE_USER_ID = 'local';

/**
 * Calculate variance of an array of numbers
 */
function calculateVariance(numbers: number[]): number {
    if (numbers.length === 0) return 0;
    const mean = numbers.reduce((a, b) => a + b, 0) / numbers.length;
    const squaredDiffs = numbers.map(x => Math.pow(x - mean, 2));
    return squaredDiffs.reduce((a, b) => a + b, 0) / numbers.length;
}

/**
 * Get daily spending for a specific date
 */
async function getDailySpent(date: Date): Promise<number> {
    const startOfDay = new Date(date.getFullYear(), date.getMonth(), date.getDate());
    const endOfDay = new Date(startOfDay);
    endOfDay.setDate(endOfDay.getDate() + 1);

    const transactions = await db.transactions
        .where('date')
        .between(startOfDay, endOfDay)
        .and(t => t.type === 'expense')
        .toArray();

    return transactions.reduce((sum, t) => sum + t.amount, 0);
}

/**
 * Build behavioral profile based on last 30 days
 */
export async function buildBehavioralProfile(): Promise<BehavioralProfile> {
    const now = new Date();
    const last30Days = new Date(now);
    last30Days.setDate(last30Days.getDate() - 30);

    const recentTransactions = await db.transactions
        .where('date')
        .above(last30Days)
        .toArray();

    // Calculate spending frequency
    const frequency = recentTransactions.length / 30;

    // Build hourly pattern
    const hourlyPattern: Record<number, number> = {};
    recentTransactions.forEach(t => {
        const hour = new Date(t.date).getHours();
        hourlyPattern[hour] = (hourlyPattern[hour] || 0) + 1;
    });

    // Analyze overruns
    let overruns = 0;
    let totalOverrun = 0;

    for (let i = 0; i < 30; i++) {
        const day = new Date(now);
        day.setDate(day.getDate() - i);
        const dailyCap = await getRecommendedDailyCap();
        const spent = await getDailySpent(day);

        if (spent > dailyCap && dailyCap > 0) {
            overruns++;
            totalOverrun += (spent - dailyCap);
        }
    }

    return {
        id: PROFILE_USER_ID,
        userId: PROFILE_USER_ID,
        spendingFrequency: frequency,
        hourlyPattern,
        overrunCount: overruns,
        averageOverrun: overruns > 0 ? totalOverrun / overruns : 0,
        lastUpdated: now,
    };
}

/**
 * Get or create behavioral profile
 */
export async function getOrCreateBehavioralProfile(): Promise<BehavioralProfile> {
    const existing = await db.behavioralProfiles.get(PROFILE_USER_ID);

    if (existing) {
        // Update if more than 24h old
        const hoursSinceUpdate = (Date.now() - new Date(existing.lastUpdated).getTime()) / (1000 * 60 * 60);
        if (hoursSinceUpdate > 24) {
            const updated = await buildBehavioralProfile();
            await db.behavioralProfiles.put(updated);
            return updated;
        }
        return existing;
    }

    // Create new profile
    const profile = await buildBehavioralProfile();
    await db.behavioralProfiles.add(profile);
    return profile;
}

/**
 * Calculate risk score (0 = disciplined, 1 = at risk)
 */
export function getRiskScore(profile: BehavioralProfile): number {
    const frequencyScore = profile.spendingFrequency > 5 ? 0.3 : 0;
    const overrunScore = profile.overrunCount > 10 ? 0.4 : (profile.overrunCount / 25);
    const eveningScore = (profile.hourlyPattern[20] || 0) > 3 ? 0.3 : 0;

    return Math.min(frequencyScore + overrunScore + eveningScore, 1.0);
}

/**
 * Get advice level based on risk score
 */
export function getAdviceLevel(profile: BehavioralProfile): 'minimal' | 'standard' | 'frequent' {
    const riskScore = getRiskScore(profile);
    if (riskScore > 0.6) return 'frequent';
    if (riskScore < 0.3) return 'minimal';
    return 'standard';
}

/**
 * Check if user tends to spend in the evening
 */
export function isEveningSpender(profile: BehavioralProfile): boolean {
    const eveningTransactions =
        (profile.hourlyPattern[18] || 0) +
        (profile.hourlyPattern[19] || 0) +
        (profile.hourlyPattern[20] || 0) +
        (profile.hourlyPattern[21] || 0);

    const totalTransactions = Object.values(profile.hourlyPattern).reduce((a, b) => a + b, 0);

    return totalTransactions > 0 && (eveningTransactions / totalTransactions) > 0.4;
}
